compiler/src/iree/compiler/Dialect/TensorExt/IR/TensorExtOps.cpp

```diff
@@ -554,6 +561,239 @@ OpFoldResult CastToRaggedShapeOp::getNumRaggedRowsAsOfr() {
+    mapping.map(op->getResult(0), dimReplacementVal);
+  }
+  SmallVector<Value> clonedQueriedValues;
+  clonedQueriedValues.reserve(queriedValues.size());
+  for (Value qv : queriedValues) {
+    Value cv = mapping.lookupOrDefault(qv);
+    clonedQueriedValues.push_back(cv);
+  }
+  return clonedQueriedValues;
+}
+
+FailureOr<SmallVector<Range>>
+CastToRaggedShapeOp::getEstimatedLoopRange(RewriterBase &rewriter,
+                                           ArrayRef<int64_t> sparseDims,
+                                           ArrayRef<Range> givenRange) {
+  SmallVector<int64_t> expectedSparseDims =
+      getResultSparseEncoding().getSparseDimensions();
+  assert(expectedSparseDims.size() == 2 &&
+         "invalid specification of op with more than two sparse dimensions");
+  if (expectedSparseDims != sparseDims) {
+    return emitOpError(
+        "cannot estimate the loop range for given sparse dimensions");
+  }
+  assert(givenRange.size() == sparseDims.size() &&
+         "expected ranges for all dims");
+
+  SmallVector<Range> estimatedRange = llvm::to_vector(givenRange);
+
+  DominanceInfo dominanceInfo(getOperation()->getParentOp());
+  Location loc = getLoc();
+
+  // For the outer sparse dimension, the estimated range is obtained by
+  // replacing `dim(%sparseOp, outerSparseDim)` with `%num_ragged_rows` in the
+  // backward slice of the upper bound.
+  Value outerUb =
+      getValueOrCreateConstantIndexOp(rewriter, loc, givenRange[0].size);
+  // Replace the `memref.dim`/`tensor.dim` operation in the backward slice of
+  // the upper bound with the number of ragged rows.
+  FailureOr<SmallVector<Value>> outerDimReplacementResult =
+      cloneAndReplaceDimInBackwardSlice(
+          rewriter, loc, dominanceInfo, outerUb, *this, expectedSparseDims[0],
+          [&](RewriterBase &rewriter, Location loc) {
+            OpFoldResult numRaggedRows = getNumRaggedRowsAsOfr();
+            Value numRaggedRowsVal =
+                getValueOrCreateConstantIndexOp(rewriter, loc, numRaggedRows);
+            return numRaggedRowsVal;
+          },
+          {outerUb});
+  if (failed(outerDimReplacementResult)) {
+    return emitOpError("failed to replace outer dim in backward slice");
+  }
+  estimatedRange[0].size = outerDimReplacementResult.value()[0];
+
+  // For the inner sparse dimension, the estimated range is obtained by
+  // replacing the `dim(%sparseOp, innerSparseDim)` with
+  // - either `%max_column_lengths` if given, or
+  // - `dim(%source, outerSparseDim) / %num_ragged_rows`.
+
+  Value innerUb =
+      getValueOrCreateConstantIndexOp(rewriter, loc, givenRange[1].size);
+  FailureOr<SmallVector<Value>> innerDimReplacementResult =
+      cloneAndReplaceDimInBackwardSlice(
+          rewriter, loc, dominanceInfo, innerUb, *this, expectedSparseDims[1],
+          [&](RewriterBase &rewriter, Location loc) {
+            if (Value maxRaggedColumnLength = getAvgRaggedColumnLength()) {
+              return maxRaggedColumnLength;
```

Comment:
The op description calls it maxRaggedColumnLength but the field is still avgRaggedColumnLength. The inconsistent naming is confusing, I'm guessing it's supposed to be maxRaggedColumnLength?